### DISCUSSION TITLE
can't use a template to make a request with the "tts" domain, but can do so just fine when the node itself is configured to, am I doing something wrong?

### DISCUSSION THREAD

**Discussion by blanknam3d** [zachowj/node-red-contrib-home-assistant-websocket#1098] [Q&A]
I feel like this might be a bug, but I also know absolutely nothing about programming anything at all - then again, I'm also heavily persistent & virtually always find a way to make stuff I'm new to work without any help.

This problem's stumped me though - I'm trying to send a tts.google_translate_say service call to HA via a Discord bot user. I'm pretty sure I have it all formatted properly - after it goes through the rest of my flow from Discord, it enters a subflow for HA, where the final msg.payload sent to the HA node is something like this:
```
{
    "domain": "tts",
    "service": "google_translate_say",
    "data": {
        "entity_id": "media_player.livingroom_camera_speaker",
        "message": "Whatever message was sent in Discord"
    }
}
```

What I expect to see, of course, is the "call service" node forwarding "success!" to a Discord node, and the camera in my living room playing the message. Instead, I'm getting a debug message: `"call service node is missing api "domain" property, not found in config or payload"`

Even more confusingly, when I set up a second node to test with, specifically configuring it to use the TTS domain and service, & set the output to just give it what was sent to HA, the output looked exactly like my template, besides small differences in formatting - just that it worked whereas my template payload did not. Furthermore, when I modified my template to exactly match the output of the call service node, it still didn't acknowledge the `tts` domain in the payload & just gave another debug message in Node-RED.

Am I doing something wrong here? Am I overlooking something small? Or have I legitimately stumbled across an odd bug, where it just won't accept the domain property `tts` from payload but will from config?

**Comment by zachowj** [comment]
Your above sample JSON should work. I would triple-check that the data going into the call-service node is what you expect it to be.

**Reply by blanknam3d** [reply]
I've checked it multiple times over - it seems the output isn't being messed with in any way, so that shouldn't be a problem

Furthermore, I use a similar template for commanding lights, where specifying domain works just fine:
```
{
    "domain": "light",
    "service": "{{power}}",
    "target": {
        "area_id": ["{{room}}"],
        "device_id": [""],
        "entity_id": ["{{entity}}"]
    }
}
```
Just that when I specify the TTS domain instead, it seems to think no domain was specified at all?

**Reply by zachowj** [reply]
Does the below flow work?

![image](https://github.com/zachowj/node-red-contrib-home-assistant-websocket/assets/37859597/4857d893-8f73-4234-ab01-270e806048d9)

```
[{"id":"8bb02d3b36b2516b","type":"inject","z":"120358abd7c22d30","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"{\"domain\":\"tts\",\"service\":\"google_translate_say\",\"data\":{\"entity_id\":\"media_player.livingroom_camera_speaker\",\"message\":\"Whatever message was sent in Discord\"}}","payloadType":"json","x":202,"y":5104,"wires":[["0d9404661d8bfd5b"]]},{"id":"0d9404661d8bfd5b","type":"api-call-service","z":"120358abd7c22d30","name":"","server":"","version":5,"debugenabled":false,"domain":"","service":"","areaId":[],"deviceId":[],"entityId":[],"data":"","dataType":"jsonata","mergeContext":"","mustacheAltTags":false,"outputProperties":[],"queue":"none","x":390,"y":5104,"wires":[[]]}]
```

**Reply by blanknam3d** [reply]
If I've understood how to use it properly, then unfortunately it didn't - I set the value in the inject node as JSON, as I at least think that's what the provided string is, though I did also try JSONata

**Reply by zachowj** [reply]
https://nodered.org/docs/user-guide/editor/workspace/import-export

You import the string into Node-RED

**Reply by blanknam3d** [reply]
Ah, my bad, not too familiar with everything in Node-RED just yet

Nonetheless, even when I do it properly (by actually importing the string lol), it still returns the same error as before.

**Reply by zachowj** [reply]
The above flow works on the latest versions. 

What version of node-red-contrib-home-assistant-websocket are you running?
The version of Home Assistant?
What is the error?

**Reply by blanknam3d** [reply]
Using Home Assistant (container) 2023.8.1, actually just updated node-red-contrib-home-assistant-websocket to 0.57.1 while trying to troubleshoot the issue

The error, as stated in the Node-RED debug messages:
"call service node is missing api "domain" property, not found in config or payload"

(accidentally posted as a comment on my own discussion, oops)

**Reply by blanknam3d** [reply]
Update - The issue turned out to be with my Home Assistant version. Decided to finally try going through the complete PITA that updating a Home Assistant container install is just to troubleshoot this. Sure enough, after spending a few hours figuring that out, my problem has been resolved!
